Allow images to be excluded from scanning

stefanprodan opened the pull request:
- add registry-exclude-image flag that accepts a list of glob expressions
- exclude Kubernetes system images by default `k8s.gcr.io/*`
- add exclusion tests
- add usage examples to docs

Fix #1654 images can be excluded by registry domain eg.:
 `--registry-exclude-image=gcr.io/*,quay.io/*`

Fix #1509 it is now possible to disable the image cache completely with:
 `--registry-exclude-image=*`

squaremo commented on cluster/kubernetes/images.go:
```diff
@@ -140,5 +141,16 @@ func (c *Cluster) ImagesToFetch() registry.ImageCreds {
 		}
 	}
 
+	// remove images based on the glob exclusion list
```
I wonder if it is possible to exit (continue) early if an image it be excluded, rather than do all the work then undo some of it. The way `mergeCredentials` works is to get all the secrets, then record them against each images used by the workload. Exiting early would mean testing whether _any_ image in the workload is to be included, and only continuing if that's the case. In other words, compile the list of images in a workload that are to be included, first, and if len==0, exit.

This adds a bit of complication to `mergeCredentials`, but not much more than it removes from here; with the benefit that it'll reduce some Kubernetes API use (no use looking up image secrets for images we don't care about).

If it turns out that it's much simpler to do it here, I'd consider putting exclusion in a separate ImageFetch wrapper, like the ECR auth bit, instead.

squaremo replied:
In any case, it'd be good to see some tests of the exclusion -- both as a guard against regressions, and as docs of how it's intended to work.

stefanprodan replied:
I've moved the exclusion logic to `mergeCredentials` and added tests.

squaremo commented on cmd/fluxd/main.go:
```diff
@@ -105,6 +105,7 @@ func main() {
 		registryBurst        = fs.Int("registry-burst", defaultRemoteConnections, "maximum number of warmer connections to remote and memcache")
 		registryTrace        = fs.Bool("registry-trace", false, "output trace of image registry requests to log")
 		registryInsecure     = fs.StringSlice("registry-insecure-host", []string{}, "use HTTP for this image registry domain (e.g., registry.cluster.local), instead of HTTPS")
+		registryExcludeImage = fs.StringSlice("registry-exclude-image", []string{"k8s.gcr.io/*"}, "Do not scan images that match these glob expressions; the default is to exclude the 'k8s.gcr.io/*' images")
```
the prevailing convention is to not capitalise the usage message

stephenmoloney commented:
just wondering, when is it desirable to not scan images in a registry?

stefanprodan commented:
@stephenmoloney see @alanjcastonguay comment on #1654

stefanprodan commented:
@squaremo should we exclude `quay.io/weaveworks/flux*` and `quay.io/weaveworks/helm-operator*` by default?

squaremo commented:
> should we exclude `quay.io/weaveworks/flux*` and `quay.io/weaveworks/helm-operator*` by default?

No; people routinely automate those (e.g., Weaveworks).

stefanprodan commented:
@squaremo I've addressed your comments in the latest commits and I've also made changes to the docs. Let me know if I've missed something. Thanks

squaremo commented on cluster/kubernetes/images.go:
```diff
@@ -142,3 +141,50 @@ func (c *Cluster) ImagesToFetch() registry.ImageCreds {
+// filterImages returns an image list from a pod spec
+// by removing those matching the exclusion list
+func (c *Cluster) filterImages(podTemplate apiv1.PodTemplateSpec) []image.Name {
+	images := []image.Name{}
```
`var images []image.Name` is very slightly more idiomatic.

squaremo commented on cluster/kubernetes/images.go:
```diff
@@ -142,3 +141,50 @@ func (c *Cluster) ImagesToFetch() registry.ImageCreds {
+		include := true
+		for _, exp := range c.imageExcludeList {
+			if glob.Glob(exp, imageName) {
+				include = false
```
There's no point in continuing if it's already false.

squaremo commented on cluster/kubernetes/images.go:
```diff
@@ -142,3 +141,50 @@ func (c *Cluster) ImagesToFetch() registry.ImageCreds {
+		}
+		if include {
+			result = append(result, imageID)
+		} else {
```
Remove this branch if it's not actually going to do anything.

squaremo commented on cluster/kubernetes/images_test.go:
```diff
@@ -66,11 +67,51 @@ func TestMergeCredentials(t *testing.T) {
 
 	creds := registry.ImageCreds{}
-	mergeCredentials(noopLog, client, ns, spec, creds, make(map[string]registry.Credentials))
+	cluster := NewCluster(clientset, nil, nil, nil, log.NewNopLogger(), []string{}, []string{})
```
Hrmm, it's a bit unwieldy that you have to create a cluster in order to filter the images. Maybe construct the filter independently of the cluster.

squaremo replied:
That way you can make it a no-op if there are no exclusions, and you can build the list of included images in one step in mergeCredentials (rather than building a list of all of them, then filtering it down).

squaremo replied:
What I meant is this: given a predicate `includeImage(imageName) bool`, you can step through a pod template, testing each image extracted against the predicate as you go. There's no need to build the list of images, then run through it again to filter it.

How do you get such a predicate? You can just construct it -- you have the algorithm right here -- and it need not depend on the cluster, since all that's doing is providing the list of exclusions.

squaremo commented on deploy/memcache-dep.yaml:
```diff
@@ -1,5 +1,5 @@
 ---
-# You can optionally deploy memcache, for the Flux daemon to cache
+# The memcached deployment is used by Flux to cache
```
This does not seem to be a related change.

stefanprodan replied:
Because of that "optionally" we've got users on Slack that thought Flux will work without memcached. But even with `--registry-exclude-image=*` memcached is mandatory.

squaremo replied:
I'm not arguing that the sentence is correct, just that you should keep the PR on point.

squaremo commented on site/faq.md:
```diff
@@ -29,11 +29,11 @@ menu_order: 60
 - [Flux Helm Operator questions](#flux-helm-operator-questions)
   * [I'm using SSL between Helm and Tiller. How can I configure Flux to use the certificate?](#im-using-ssl-between-helm-and-tiller-how-can-i-configure-flux-to-use-the-certificate)
-  * [I've deleted a FluxHelmRelease file from Git. Why is the Helm release still running on my cluster?](#ive-deleted-a-fluxhelmrelease-file-from-git-why-is-the-helm-release-still-running-on-my-cluster)
+  * [I've deleted a HelmRelease file from Git. Why is the Helm release still running on my cluster?](#ive-deleted-a-helmrelease-file-from-git-why-is-the-helm-release-still-running-on-my-cluster)
```
Unrelated change. Not unwelcome, but belongs in another PR.

stefanprodan commented:
@squaremo I've removed the docs changes from this PR and I've addressed your comments in the latest commit. As for making the tests work without cluster I'm not really sure how to do that without adding the exclusion list on all functions... I would leave that for another PR if it's ok with you.

squaremo commented on cluster/kubernetes/images.go:
```diff
@@ -14,7 +15,18 @@ import (
+	imageCreds registry.ImageCreds,
+	seenCreds map[string]registry.Credentials) {
+	// filter the images based on the exclusion list
+	images := filterImages(podTemplate)
```
All that happens to `filterImages` is that it's applied to one of the other arguments -- you may as well supply the result.

squaremo commented:
That last commit is a nice refinement :+1:
